WiFiManager.cpp

```diff
@@ -136,24 +168,44 @@ boolean WiFiManager::autoConnect(char const *apName, char const *apPassword) {
   //String ssid = getSSID();
   //String pass = getPassword();
 
-  // attempt to connect; should it fail, fall back to AP
+  // device will attempt to connect by itself; wait 10 secs
+  // to see if it succeeds and should it fail, fall back to AP
   WiFi.mode(WIFI_STA);
-
-  if (connectWifi("", "") == WL_CONNECTED)   {
-    DEBUG_WM(F("IP Address:"));
-    DEBUG_WM(WiFi.localIP());
-    //connected
-    return true;
-  }
+  unsigned long startedAt = millis();
+    while(millis() - startedAt < 10000)
+    {
+        delay(100);
+        if (WiFi.status()==WL_CONNECTED) {
+			float waited = (millis()- startedAt);
+			DEBUG_WM(F("After waiting "));
+			DEBUG_WM(waited/1000);
+			DEBUG_WM(F(" secs local ip: "));
+			DEBUG_WM(WiFi.localIP());
+            return true;
+		}
+    }
 
   return startConfigPortal(apName, apPassword);
 }
 
+boolean  WiFiManager::startConfigPortal() {
+  String ssid = "ESP" + String(ESP.getChipId());
+  return startConfigPortal(ssid.c_str());
+}
+
 boolean  WiFiManager::startConfigPortal(char const *apName, char const *apPassword) {
   //setup AP
-  WiFi.mode(WIFI_AP_STA);
-  DEBUG_WM("SET AP STA");
-
+  int connRes = WiFi.waitForConnectResult();
+  if (connRes == WL_CONNECTED){
+	  WiFi.mode(WIFI_AP_STA); //Dual mode works fine if it is connected to WiFi
+	  DEBUG_WM("SET AP STA");
+  	}
+  	else {
+    WiFi.mode(WIFI_AP); // Dual mode becomes flaky if not connected to a WiFi network.
```

Comment:
Yes, apart from the issue you highlighted, switching channels causes confusion elsewhere. When scanning for WiFi networks the ESP drops the WiFi network to scan the other channels. The browsing device sometimes connects to a different WiFi network automatically before the AP WiFi is restarted. Users don't notice the network change and it confuses them. This is more likely when there is a high density of WiFi networks. The AP WiFi network also disappears when it needs to change channels to connect to a WiFi network and reappears on the new channel. Browsing devices will often switch WiFi networks when this occurs, again unnoticed and causing confusion.

The scanning issue can be improved by scanning before setting up the WiFi network and then listing networks from memory rather than re-scanning. 

It is also why an app is easier to use than a browser because, unlike a browser, an app can control  which network it connects to and switch back and forth as required.